### DISCUSSION TITLE
Autoformatting, type-checking, and linting

### DISCUSSION THREAD

**Discussion by impredicative** [astral-sh/rye#592] [General]
In the Python world, auto-formatting, type-checking, and linting are three separate concerns. They are not one and the same.

Regarding auto-formatting, I tried `ruff` via `rye lint` but it does **not** look to have sane defaults, making it useless as I see it. It did not detect or fix various simple issues. I find `black` more suitable.

Regarding type-checking, `mypy` is critical to me, although the choice of tool may vary.

Regarding any leftover linting, `pylint` is what I would compare any other Python linter with.

---

My point is that `rye lint` should probably be removed unless all of the above can be tackled.

**Comment by mitsuhiko** [comment]
You would need to substantiate the claims a bit that `ruff` is "useless". Can you demonstrate a case where you expect the lint to detect something but it did not? I find it vastly more likely that there is a bug and we accidentally skipped over a file. Yes, type checking is something that definitely should be added but it's not yet fully clear to me how to best approach this given that `mypy` has to be sitting inside the dependency tree. There might also be a better choice than `mypy` and it's something I need to investigate.

**Comment by impredicative** [comment]
I did check `rye lint -v` (verbose option) output to confirm that the file was not skipped by `ruff`. It didn't even check or fix basic whitespace issues, although it did find one other issue. More later.

**Comment by impredicative** [comment]
With the default options, `rye lint --fix` (via `ruff`) doesn't strip multiple trailing newlines in a file. It also doesn't normalize string quotes (single-quoted strings versus double-quoted strings), whereas `black` does. These are common auto-formatting issues.

**Reply by cnpryer** [reply]
Have you tried the formatter? What your describing sounds like it'd be handled by the formatter, not the linter. The formatter should also provide options for configuring quote preference.

**Reply by impredicative** [reply]
@cnpryer `rye fmt` addressed it. What had me confused was that `rye lint --fix` is in a manner of speaking also doing some formatting. As such there are two formatters.

**Reply by cnpryer** [reply]
If there isn't already, there will be a `ruff check` command that consolidates these features in a way. Maybe that's what you're expecting?

I wouldn't equate formatting to fixing lints.

**Reply by impredicative** [reply]
@cnpryer I added this to my `pyproject.toml`:

```toml
[tool.rye.scripts]  # Run as "rye run <command>"
check = { chain = ["check-lint", "check-fmt" ] }
"check-lint" = "rye lint"
"check-fmt" = "rye fmt --check"
fix = { chain = ["fix-lint", "fix-fmt" ] }
"fix-lint" = "rye lint --fix"
"fix-fmt" = "rye fmt"
```

**Comment by acalvino4** [comment]
+1 for integrating a typechecker into the 'hassle-free' python toolchain. don't particularly care if it's mypy or something else, as long as it works and is fast!

**Comment by impredicative** [comment]
It sort of does matter if it's mypy or something else. Python's `typing` module has been updating in every release of Python, and various type checkers that are not mypy could struggle to keep up.

The same concern also applies with various linters and formatters that may struggle to keep up with newer Python syntax.

**Comment by mitsuhiko** [comment]
The problem with adding typing support today is that the choices of type checkers are not great. I think the best type checker available today for multiple reasons is `pyright` but unfortunately it also requires `node` to be installed. It also would require a lot of custom code around it to make it work well for no LSP kind of setups. I do think that at one point it would be great though to have the type checker also integrated.

If nothing better than `pyright` comes up, I would honestly consider transpiling `pyright` or bundling a javascript interpreter optionally to run this thing.

**Reply by impredicative** [reply]
I don't know if https://github.com/Distributive-Network/PythonMonkey qualifies as a JS interpreter.

**Reply by impredicative** [reply]
I guess it now stands to reason that Astral is working on a Python type-checker.

**Comment by lgc2333** [comment]
Maybe [basedpyright](https://github.com/DetachHead/basedpyright) is an another good choice for type checking? It can be installed via pip (it also requires node but the node binary will install via pip too)
